### PR TITLE
[WIP] Add command: `bqetl query schema`

### PR DESCRIPTION
**Pull request by jklukas**
A potential alternative to https://github.com/mozilla/bigquery-etl/issues/1539

With this functionality, adding a new field to clients_daily would look like
the following:

- File a PR that updates the clients_daily_v6 query.sql
- Get review (some tests are still failing)
- When ready to merge, run:

```bash
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_daily_event_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_last_seen_event_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_daily_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_last_seen_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_first_seen_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_daily_joined_v1
bqetl query schema --update moz-fx-data-shared-prod.telemetry_derived.clients_last_seen_joined_v1
```

- All tests should now pass
- Merge the PR
- Incremental queries should all now run successfully.

This is currently marked as a draft. I'm seeking feedback about the general approach here before deciding on whether to fully flesh it out and document the workflow for applying schema updates to the clients_daily ETL stream.

### CODE REVIEW

File: bigquery_etl/cli/query.py [hunk 1]
```diff
@@ -462,6 +463,66 @@ def backfill(ctx, name, sql_dir, project_id, start_date, end_date, exclude, dry_
...
+        config = bigquery.QueryJobConfig(
+            dry_run=True,
+            default_dataset=f"{project}.{dataset}",
+            query_parameters=[
+                bigquery.ScalarQueryParameter(
+                    "submission_date", "DATE", "2000-01-01"
+                ),
+            ]
+        )
+        new_schema = client.query(sql, config)._job_statistics()['schema']
```
Comment (jklukas) [inline bigquery_etl/cli/query.py:520]: Note that this relies on a non-public `_job_statistics` method to get at the schema part of the response.

**Comment by scholtzan** [comment]
I do like that this option builds on the existing bigquery-etl structure by using `init.sql` files. Handling of errors, for example, the `init.sql` changes are incompatible with the existing schema etc. will be necessary.

Long-term I do think that replacing the `init.sql` files by schema files would be advantageous (most of the reasons have been mentioned in https://github.com/mozilla/bigquery-etl/issues/1539: column descriptions, schema files could be autogenerated from queries, schema files could be generated for existing tables, ...). From what I gathered, the main reason supporting schema files might not be ideal is the complexity added to the workflow making it less user friendly. The workflow proposed here seems to be almost identical, but less powerful, compared to schema files. `init.sql` also might imply to users that it is used only for initializing the table and not for ongoing table schema updates.

**Comment by jklukas** [comment]
> From what I gathered, the main reason supporting schema files might not be ideal is the complexity added to the workflow making it less user friendly. The workflow proposed here seems to be almost identical, but less powerful, compared to schema files. init.sql also might imply to users that it is used only for initializing the table and not for ongoing table schema updates.

These are good points. Part of what I found problematic about checking in schema files is that modifying them by hand could be problematic. So perhaps we need some combination of schemas checked in and tooling to maintain and apply them. We'd probably need to maintain logic for diffing the schemas intelligently, though, which would be significantly more to maintain than what we have here.

**Comment by jklukas** [comment]
Superseded by #1749

**Pull request closed without merging**